File: packages/fabric/lib/template.ts
import { z } from 'zod';
import {
  fetchClient,
  clientToTemplate,
  buildClasspath,
  buildLaunch,
} from '@opys/minecraft-vanilla';
import {
  type Artifact,
  type ValDefs,
  type Launch,
  type Val,
  type Valset,
  sourceUrl,
  fetchWithRetry,
} from '@opys/core';
import { mergeArgs, parseArguments, parseMaven } from '@opys/mojang';
import type { MavenCoord } from '@opys/mojang';
import {
  resolveFabricVersion,
  DEFAULT_FABRIC_META,
  type FabricRelease,
} from './resolver';

export interface FabricOptions {
  /** Minecraft (game) version, e.g. `1.21.4`. */
  version: string;
  /**
   * Fabric loader version, e.g. `0.16.10`. Omit for the latest stable loader
   * build that targets `version`.
   */
  loader?: string;
  /** Fabric Meta base URL. Default: `https://meta.fabricmc.net`. */
  source?: string;
}

export interface FabricTemplate {
  /** Vanilla MC artifacts + Fabric loader/intermediary library artifacts. */
  artifacts: Artifact[];
  vars: ValDefs;
  /** Assembled Launch — drop straight into `manifest.launch`. */
  launch: Launch;
  /** JVM args alone, for composition (e.g. interleaving authliberty's `-javaagent`). */
  jvmArgs: Valset;
  /** Main class wrapped as a `Val` so it spreads into a `Valset`. */
  mainClass: Val;
  /** Game args alone, for composition. */
  gameArgs: Valset;
}

/**
 * A Fabric launcher-profile library: a Maven coordinate plus the repo base it
 * lives in. Newer Meta responses also ship per-file `sha1`/`size` — used for
 * artifact integrity when present.
 */
const FabricLibrarySchema = z.object({
  name: z.string(),
  url: z.string(),
  sha1: z.string().optional(),
  size: z.number().optional(),
});

const FabricProfileSchema = z.object({
  inheritsFrom: z.string(),
  mainClass: z.string(),
  arguments: z.unknown().optional(),
  libraries: z.array(FabricLibrarySchema),
});

type FabricLibrary = z.infer<typeof FabricLibrarySchema>;

/** Maven coord with `version` required. */
type ResolvedCoord = MavenCoord & { version: string };

/**
 * Build a maven path from a coord:
 * `<group_path>/<artifact>/<version>/<artifact>-<version>[-<classifier>].<ext>`.
 */
function mavenPath(c: ResolvedCoord): string {
  const groupPath = c.groupId.replace(/\./g, '/');
  const ext = c.packaging ?? 'jar';
  const filename = c.classifier
    ? `${c.artifactId}-${c.version}-${c.classifier}.${ext}`
    : `${c.artifactId}-${c.version}.${ext}`;
  return `${groupPath}/${c.artifactId}/${c.version}/${filename}`;
}

function requireVersioned(name: string): ResolvedCoord {
  const coord = parseMaven(name);
  if (coord.version === undefined) {
    throw new Error(`Fabric library coordinate '${name}' has no version`);
  }
  return coord as ResolvedCoord;
}

/** A Fabric profile library → artifact + its `${library_directory}`-relative path. */
function fabricLibArtifact(lib: FabricLibrary): {
  artifact: Artifact;
  path: string;
} {
  const path = mavenPath(requireVersioned(lib.name));
  const url = `${lib.url.replace(/\/+$/, '')}/${path}`;
  const artifact: Artifact = {
    path: `\${library_directory}/${path}`,
    source: sourceUrl(url),
    rules: [],
    ...(lib.size ? { size: lib.size } : {}),
    ...(lib.sha1 ? { integrity: { sha1: lib.sha1 } } : {}),
  };
  return { artifact, path };
}

async function fetchProfile(
  url: string,
): Promise<z.infer<typeof FabricProfileSchema>> {
  const res = await fetchWithRetry(url);
  if (!res.ok) {
    throw new Error(
      `Failed to download Fabric profile from ${url}: ${res.status} ${res.statusText}`,
    );
  }
  return FabricProfileSchema.parse(await res.json());
}

export async function resolveFabric(
  options: FabricOptions,
): Promise<FabricTemplate> {
  const meta = options.source ?? DEFAULT_FABRIC_META;
  const release: FabricRelease = await resolveFabricVersion(
    options.version,
    meta,
    options.loader,
  );

  const profile = await fetchProfile(release.profileUrl);

  const { client } = await fetchClient(profile.inheritsFrom);
  const mc = await clientToTemplate(client);

  // Fabric profile libraries are Maven coord + repo base, no OS rules and no
  // natives — every entry goes on the classpath unconditionally.
  const fabricLibs = profile.libraries.map(fabricLibArtifact);

  const libPaths = [
    ...client.libraries.map((l) => ({
      rules: l.rules,
      artifactPath: `\${library_directory}/${l.artifact.path}`,
    })),
    ...fabricLibs.map(({ path }) => ({
      rules: [],
      artifactPath: `\${library_directory}/${path}`,
    })),
  ];

  const classpath = buildClasspath(libPaths, '${version_dir}/client.jar');

  const fabricArgs = parseArguments(profile.arguments ?? { game: [], jvm: [] });
  const merged = mergeArgs(client.args, fabricArgs);

  const vars: ValDefs = { ...mc.vars, classpath };
  const parts = buildLaunch(profile.mainClass, merged.game, merged.jvm);

  return {
    artifacts: [...mc.artifacts, ...fabricLibs.map(({ artifact }) => artifact)],
    vars,
    ...parts,
  };
}

File: crates/opys-fabric/src/template.rs
//! The `fabric` plugin: a Fabric launcher profile layered onto vanilla.

use std::collections::HashSet;

use opys_core::{Artifact, ConditionalVal, Launch, Val, ValDef, ValDefs};
use opys_dev::{Contribution, LaunchFragment, PluginOutput};
use opys_minecraft_vanilla::{
    build_launch, fetch_client, inherited_classpath, resolve_client_template, superseded,
    ClasspathEntry, MinecraftTemplate,
};
use opys_mojang::Client;
use serde::{Deserialize, Serialize};

use crate::error::FabricError;
use crate::profile::{fetch_profile, library_artifact, FabricProfile};
use crate::resolver::{resolve_fabric_version, DEFAULT_FABRIC_META};

/// The name this plugin claims in the plugin map.
pub const PLUGIN_NAME: &str = "fabric";

/// What to resolve.
#[derive(Debug, Clone, Default, PartialEq, Eq, Serialize, Deserialize)]
#[serde(default, rename_all = "camelCase")]
pub struct FabricOptions {
    /// Minecraft (game) version, e.g. `1.21.4`.
    pub version: String,
    /// Fabric loader version, e.g. `0.16.10`. `None` takes the latest stable
    /// loader build targeting `version`.
    pub loader: Option<String>,
    /// Fabric Meta base URL. `None` is [`DEFAULT_FABRIC_META`].
    pub source: Option<String>,
    /// Overrides the canonical Mojang version-manifest URL — a mirror, or the
    /// seam the tests point at a loopback server.
    pub manifest_base: Option<String>,
}

/// Everything a Fabric profile plus its vanilla base contributes to a manifest.
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct FabricTemplate {
    /// Vanilla artifacts followed by the loader's own libraries.
    pub artifacts: Vec<Artifact>,
    pub vars: ValDefs,
    /// Per-OS classpath arms (also baked into `vars.classpath`), exposed so a
    /// plugin stacked on top of Fabric can rebuild it with its own libraries.
    pub classpath: Vec<ConditionalVal>,
    /// Assembled launch — drop straight into `manifest.launch`.
    pub launch: Launch,
    /// JVM args alone, for composition (e.g. interleaving an auth `-javaagent`).
    pub jvm_args: Vec<Val>,
    /// Main class wrapped as a `Val` so it spreads into a `Valset`.
    pub main_class: Val,
    /// Game args alone, for composition.
    pub game_args: Vec<Val>,
}

/// Resolve Fabric: pick the loader build, read its profile, fetch the vanilla
/// version it inherits from, then fold the two together.
pub fn resolve_fabric(options: &FabricOptions) -> Result<FabricTemplate, FabricError> {
    let meta = options.source.as_deref().unwrap_or(DEFAULT_FABRIC_META);
    let release = resolve_fabric_version(&options.version, meta, options.loader.as_deref())?;
    let profile = fetch_profile(&release.profile_url)?;

    let (_, client) = fetch_client(
        Some(&profile.inherits_from),
        options.manifest_base.as_deref(),
    )?;
    let vanilla = resolve_client_template(&client)?;

    profile_to_template(&profile, &client, &vanilla)
}

/// The pure half: a profile, the vanilla version JSON it inherits from, and
/// that version's already-mapped template, in; a Fabric template out.
///
/// Split out for the same reason vanilla's `client_to_template` is: nothing
/// here touches the network, so the fold that actually defines Fabric — its
/// libraries appended to the classpath, its argument delta merged onto
/// vanilla's — is testable without a server.
pub fn profile_to_template(
    profile: &FabricProfile,
    client: &Client,
    vanilla: &MinecraftTemplate,
) -> Result<FabricTemplate, FabricError> {
    let libs: Vec<(Artifact, ClasspathEntry)> = profile
        .libraries
        .iter()
        .map(library_artifact)
        .collect::<Result<_, _>>()?;

    // A profile is an `inheritsFrom` patch, so its libraries go ahead of the
    // base version's, and a base library it supersedes drops out entirely —
    // Fabric ships its own ASM build and means it to be the only one.
    let patch: Vec<ClasspathEntry> = libs.iter().map(|(_, entry)| entry.clone()).collect();
    let base: Vec<ClasspathEntry> = client.libraries.iter().map(ClasspathEntry::of).collect();
    let classpath = inherited_classpath(&patch, &base, "${version_dir}/client.jar")?;

    let merged = client.args.merge(&profile.arguments);
    let parts = build_launch(&profile.main_class, &merged.game, &merged.jvm);

    let mut vars = vanilla.vars.clone();
    vars.insert("classpath".to_owned(), ValDef::Arms(classpath.clone()));

    // The download set follows the classpath: a superseded base library is no
    // longer on `-cp`, so fetching it would be work spent on a file nothing
    // opens.
    let dropped: HashSet<String> = superseded(&patch, &base).into_iter().collect();
    let mut artifacts: Vec<Artifact> = vanilla
        .artifacts
        .iter()
        .filter(|a| !dropped.contains(&a.path))
        .cloned()
        .collect();
    artifacts.extend(libs.into_iter().map(|(artifact, _)| artifact));

    Ok(FabricTemplate {
        artifacts,
        vars,
        classpath,
        launch: parts.launch,
        jvm_args: parts.jvm_args,
        main_class: parts.main_class,
        game_args: parts.game_args,
    })
}

/// A resolved template as a ready-to-merge contribution.
pub fn build_fabric(options: &FabricOptions) -> Result<PluginOutput, FabricError> {
    let template = resolve_fabric(options)?;
    Ok(PluginOutput {
        name: PLUGIN_NAME.to_owned(),
        contribution: Contribution {
            artifacts: template.artifacts,
            vars: template.vars,
            launch: [
                (
                    "command".to_owned(),
                    LaunchFragment::Text(template.launch.command.clone()),
                ),
                (
                    "jvmArgs".to_owned(),
                    LaunchFragment::Many(template.jvm_args),
                ),
                (
                    "mainClass".to_owned(),
                    LaunchFragment::One(template.main_class),
                ),
                (
                    "gameArgs".to_owned(),
                    LaunchFragment::Many(template.game_args),
                ),
            ]
            .into_iter()
            .collect(),
            envs: Default::default(),
        },
    })
}
